loopy/schedule/tools.py

```diff
@@ -51,10 +52,18 @@ def get_block_boundaries(schedule):
 def temporaries_read_in_subkernel(kernel, subkernel):
     from loopy.kernel.tools import get_subkernel_to_insn_id_map
     insn_ids = get_subkernel_to_insn_id_map(kernel)[subkernel]
-    return frozenset(tv
-            for insn_id in insn_ids
-            for tv in kernel.id_to_insn[insn_id].read_dependency_names()
-            if tv in kernel.temporary_variables)
+    inames = frozenset().union(*(kernel.insn_inames(insn_id)
+                                 for insn_id in insn_ids))
+    domain_idxs = {kernel.get_home_domain_index(iname) for iname in inames}
+    params = frozenset().union(*(
+        kernel.domains[dom_idx].get_var_names(isl.dim_type.param)
+        for dom_idx in domain_idxs))
+
+    return (frozenset(tv
+                      for insn_id in insn_ids
+                      for tv in kernel.id_to_insn[insn_id].read_dependency_names()
+                      if tv in kernel.temporary_variables)
```

Comment:
This'll be more efficient as unioning all the read-deps and then intersecting, maybe?